mget/request.go

```diff
@@ -0,0 +1,113 @@
+package os2
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"net/http"
+
+	"github.com/opensearch-project/opensearch-go/v2"
+	"github.com/opensearch-project/opensearch-go/v2/opensearchapi"
+
+	"github.com/CrowdStrike/opensearchtools"
+)
+
+// Request wraps the functionality of opensearchapi.MgetRequest to support a dynamic building of the request body.
+// Minimalistic to cover desired functionality and not full functionality.
+type Request[K opensearchtools.RoutableDoc] struct {
+	// Index destination for entire request
+	// if used individual documents don't need to specify the index
+	Index string `json:"-"`
+
+	Docs []Doc `json:"docs"`
+}
+
+// Add any number of docs to the Request
+func (m *Request[K]) Add(docs ...K) {
+	converted := make([]Doc, len(docs))
+	for i, doc := range docs {
+		converted[i] = Doc{
+			ID:    doc.GetID(),
+			Index: doc.RouteToIndex(),
+		}
+	}
+
+	m.Docs = append(m.Docs, converted...)
+}
+
+// Do executes the Multi-Get Request using the provided opensearch.Client
+func (m *Request[K]) Do(ctx context.Context, client *opensearch.Client) (*Response[K], error) {
+	bodyBytes, jErr := json.Marshal(m)
+	if jErr != nil {
+		return nil, jErr
+	}
+
+	osResp, rErr := opensearchapi.MgetRequest{
+		Index: m.Index,
+		Body:  bytes.NewReader(bodyBytes),
+	}.Do(ctx, client)
+
+	if rErr != nil {
+		return nil, rErr
+	}
+
+	var respBuf bytes.Buffer
+	if _, err := respBuf.ReadFrom(osResp.Body); err != nil {
+		return nil, err
+	}
+
+	t := &ResponseBody{}
+
+	if err := json.Unmarshal(respBuf.Bytes(), &t); err != nil {
+		return nil, err
+	}
+
+	var docs []GetResult[K]
+	for _, rawDoc := range t.Docs {
+		doc := GetResult[K]{}
+
+		if err := json.Unmarshal(rawDoc, &doc); err != nil {
+			doc.Error = err
+		}
+
+		docs = append(docs, doc)
+	}
+
+	resp := &Response[K]{
+		StatusCode: osResp.StatusCode,
+		Header:     osResp.Header,
+		Docs:       docs,
+	}
+
+	return resp, nil
+}
+
+// Doc represents individual document being requested in the multi-get
+type Doc struct {
```

Comment:
Is this specific to mget?  If not, probably best to define it in the root package.  Same thing applies to `Response`, `ResponseBody`, etc.

Reply:
It _could_ be re-leveraged by a basic [Get request ](https://opensearch.org/docs/latest/api-reference/document-apis/get-documents/). But that has all of the parameters as query parameters rather than a json body. So this really is specific to the MGet.

The response objects might be more reusable. Let me do a bit of digging.

Reply:
Yea, so the response structs are more specific to MGet. The StatusCode and Header are common to all requests that I'm pulling from `opensearchapi.Response`. I could extract those two as a common struct and extend it in the `mget.Response`, but I don't know if that's worth it.